### DISCUSSION TITLE
Bottom-up model consistently doesn't place/connect all points, but placed points are accurate

### DISCUSSION THREAD

**Discussion by ksalwass** [talmolab/sleap#2668] [Help!] [answered]
## Description
After a significant amount of training, it seems that my model will not continue to improve and we are looking for any advice (especially since this project is on a very tight timeline). The model is consistently accurate with the points that it does place and the validation images that pop up during training seem to suggest that it is learning all the points, but then in prediction it only places some of them, and/or it doesn't connect the neck and nose nodes. It seems like an issue with a confidence threshold? But our sigma is already relatively large so we're not sure of the best way to improve further. 

## Screenshots
Here is what the skeleton should look like:
<img width="1262" height="718" alt="image (2)" src="https://github.com/user-attachments/assets/9a1ef8a2-1b54-42e2-ae6d-7cf189de9e07" />

Here is what it returns with no max instances:
<img width="1253" height="657" alt="image (3)" src="https://github.com/user-attachments/assets/a52e8aa5-4198-42f4-a2d5-33d3f420f9d8" />

When we set max 2 instances, we get this:
<img width="1250" height="717" alt="image" src="https://github.com/user-attachments/assets/a8b76dc8-8f74-4da4-849a-156658b87f6d" />

Here is a screenshot of our current training settings:
<img width="766" height="732" alt="image (1)" src="https://github.com/user-attachments/assets/ca5b4300-6b34-40d4-ada5-69fa88ec1bd9" />

<details>
<summary>Output from sleap doctor</summary>
SLEAP System Diagnostics
========================
Generated: 2026-04-10 13:05:45

[Platform]
  OS:        Windows 10
  Platform:  Windows-10-10.0.19045-SP0
  Machine:   AMD64
  Processor: AMD64 Family 26 Model 68 Stepping 0, AuthenticAMD
  RAM:       23.5 GB / 125.7 GB
  Disk:      1.1 TB / 3.1 TB

[Python]
  Version:     3.13.9
  Executable:  C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\python.exe
  Prefix:      C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313

[UV]
  Version:         uv 0.10.2 (a788db7e5 2026-02-10)
  Path:            C:\Users\SylwestrakGPU\.local\bin\uv.EXE
  Cache Dir:       C:\Users\SylwestrakGPU\AppData\Local\uv\cache
  Tool Dir:        C:\Users\SylwestrakGPU\AppData\Roaming\uv\tools
  Tool Bin Dir:    C:\Users\SylwestrakGPU\.local\bin
  Python Dir:      C:\Users\SylwestrakGPU\AppData\Roaming\uv\python

[UV Config]
  Default Python:    (not configured)
  Resolved Python:   C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\python.exe
  Python Preference: managed (default)
  Resolution:        highest (default)
  Index Strategy:    first-index (default)
  Prerelease:        if-necessary (default)

[Conda]
  Status:             ACTIVE
  Environment:        sleap_env
  Prefix:             C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env

[GPU / CUDA]
  NVIDIA Driver: 581.57 (CUDA 13.0)
  GPU 0:         NVIDIA GeForce RTX 5070 Ti (14409 MB free / 16303 MB)
  PyTorch:       v2.10.0+cu128 (CUDA 12.8)

[Packages]
  sleap:                  v1.6.2 (pip)
  sleap-io:               v0.6.5 (pip)
  sleap-nn:               v0.1.0 (editable)
                          Location: /C:/Users/SylwestrakGPU/Documents/GitHub/sleap-nn
  numpy:                  v2.4.2 (pip)
  h5py:                   v3.15.1 (pip)
  PySide6:                v6.10.2 (pip)
  opencv-python:          v4.13.0.92 (pip)
  opencv-python-headless: v4.13.0.92 (pip)
  torch:                  v2.10.0+cu128 (pip)
  imageio:                v2.34.1 (pip)
  imageio-ffmpeg:         v0.5.1 (pip)
  av:                     v15.1.0 (pip)

[CLI Binaries]
  sleap:
    Path:      C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap.EXE
    Real Path: C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap.exe
    Source:    pip
  sleap-nn:
    Path:      C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap-nn.EXE
    Real Path: C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap-nn.exe
    Source:    pip
  sleap-nn-track:
    Path:      C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap-nn-track.EXE
    Real Path: C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sleap-nn-track.exe
    Source:    pip
  sio:
    Path:      C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sio.EXE
    Real Path: C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts\sio.exe
    Source:    pip
  ffmpeg:
    Path:      C:\ProgramData\chocolatey\bin\ffmpeg.EXE
    Real Path: C:\ProgramData\chocolatey\bin\ffmpeg.exe
    Source:    PATH
    Python:    v8.[0-essentials_build-www.gyan.dev](http://0-essentials_build-www.gyan.dev/)
  ffmpeg (imageio):
    Path:
C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Lib\site-packages\imageio_ffmpeg\binaries\ffmpeg-win64-v
4.2.2.exe
    Source:    imageio-ffmpeg
    Python:    v4.2.2

[PATH Conflicts]
  WARNING: Conda paths appear before uv paths in PATH - conda binaries may take precedence

[PATH (relevant entries)]
  C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Lib\site-packages\cv2\../../x64/vc17/bin
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env\Library\mingw-w64\bin
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env\Library\usr\bin
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env\Library\bin
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env\Scripts
  C:\Users\SylwestrakGPU\miniforge3\envs\sleap_env\bin
  C:\Users\SylwestrakGPU\miniforge3\condabin
  C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313\Scripts
  C:\Users\SylwestrakGPU\AppData\Local\Programs\Python\Python313
  C:\Users\SylwestrakGPU\.local\bin
  C:\Users\Sylwestrak GPU\.local\bin
Message Leah Blankenship
<details>

**Comment by tom21100227** [comment, marked as answer]
Hi @ksalwass , 

How many images (labelled instances) are in your training dataset? bottom-up models are very sensitive to skeleton configuration so i would recommend changing the skeletons to make it shallow-er (e.g. instead of connecting neck to nose, then nose to ears, connect nose and both ears directly to neck). 

I would recommend changing the skeleton configuration to this (rest assured that changing skeletons does not delete your existing labels): 

Connect everything anterior to neck, neck connect to com, com to every node posterior (including tails). 

Then retrain the model. 

https://legacy.sleap.ai/guides/skeletons.html

Hope this helps, 
Tom
